File: competitive_programming/leetcode/cpp/301_remove_invalid_parentheses.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

bool isValid(string s) {
    int count = 0;
    for (char c : s) {
        if (c == '(') {
            ++count;
        } else if (c == ')') {
            --count;
        }
        if (count < 0) {
            return false;
        }
    }
    return count == 0;
}

void dfs(int start, int l, int r, string s, vector<string>& ans) {
    if (l == 0 and r == 0) {
        if (isValid(s)) {
            ans.push_back(s);
        }
        return;
    }

    for (int i = start; i < s.size(); ++i) {
        if (i > start and s[i] == s[i - 1]) {
            continue;
        }
        if (s[i] == '(' || s[i] == ')') {
            string copy = s;
            copy.erase(i, 1);
            if (s[i] == '(' && l > 0) {
                dfs(i, l - 1, r, copy, ans);
            } else if (s[i] == ')' && r > 0) {
                dfs(i, l, r - 1, copy, ans);
            }
        }
    }
}

vector<string> removeInvalidParentheses(string s) {
    vector<string> ans;
    // number of parentheses to remove 
    int l = 0;
    int r = 0;
    for (char c : s) {
        l += c == '(';
        if (l == 0) {
            r += c == ')';
        } else {
            l -= c == ')';
        }
    }
    dfs(0, l, r, s, ans);
    return ans;
}

void print(vector<string> ans) {
    for (string s : ans) {
        cout << s << " ";
    }
    cout << "\n";
}

int main() {
    vector<string> ans = removeInvalidParentheses("()())()");
    print(ans);
    ans = removeInvalidParentheses("(a)())()");
    print(ans);
    ans = removeInvalidParentheses(")(");
    print(ans);
}

File: competitive_programming/leetcode/python/301_remove_invalid_parentheses.py
def dummyDfs(s):
    def dfs(start, l, r, target, stack, ans, s):
        if r > l:
            return
        if target == 0:
            if l == r:
                ans.append(stack)
            return

        for i in range(start, len(s)):
            # pruning, skip duplicate
            if i > start and s[i] == s[i - 1]:
                continue       
            stack += s[i]
            dl, dr = 0, 0
            if s[i] == '(':
                dl = 1
            elif s[i] == ')':
                dr = 1
            dfs(i + 1, l + dl, r + dr, target - 1, stack, ans, s)
            stack = stack[:-1]

    # remove all '(' at the end of string
    end = len(s)
    tail = ""
    while end >= 1 and s[end - 1] != ')':
        if s[end - 1].isalpha():
            tail = s[end - 1] + tail
        end -= 1
    s = s[:end] + tail
    if len(s) == 0:
        return [""]

    # calculate the least number of '(' or ')' has to be removed 
    l, r = 0, 0
    least = 0
    for c in s:
        if c == '(':
            l += 1
        elif c == ')':
            r += 1
        if r > l:
            least += 1
            r -= 1
    least += l - r
            
    ans = []
    start, l, r = 0, 0, 0
    dfs(start, l, r, len(s) - least, "", ans, s)
    return ans

print(dummyDfs("()())()"))
print(dummyDfs("(a)())()"))
print(dummyDfs(")("))
print(dummyDfs("(((("))
print(dummyDfs("))"))
print(dummyDfs(""))
print(dummyDfs("n"))
print(dummyDfs("(()"))